src/vs/workbench/contrib/terminal/browser/terminalInstance.ts

```diff
@@ -1058,6 +1060,47 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 		this._terminalAltBufferActiveContextKey.set(!!(this.xterm && this.xterm.raw.buffer.active === this.xterm.raw.buffer.alternate));
 	}
 
+	private async _shouldPasteText(text: string): Promise<boolean> {
+		let textForLines = text.split(/(\r\n|\n|\r)/);
+		let confirmation: IConfirmationResult;
+
+		// If the clipboard has only one line, no prompt will be triggered
+		if (textForLines.length === 1 || this._configurationService.getValue<boolean>(TerminalSettingId.MultiLinePasteWarning) === false) {
+			confirmation = { confirmed: true };
+		} else {
+			let message = nls.localize('confirmMoveTrashMessageFilesAndDirectories', "Are you sure you want to paste the following {0} lines to the terminal?", text.split(/\r\n|\n|\r/).length);
+
+			let displayItemsCount = 3;
+			let detail = textForLines.slice(0, displayItemsCount).join('');
+			if (textForLines.length > displayItemsCount) {
+				detail += '...';
+			}
+			let primaryButton = nls.localize({ key: 'multiLinePasteButton', comment: ['&& denotes a mnemonic'] }, "&&Paste");
+
+			confirmation = await this._dialogService.confirm({
+				type: 'question',
+				message: message,
+				detail: detail,
+				primaryButton: primaryButton,
+				checkbox: {
+					label: nls.localize('doNotAskAgain', "Do not ask me again")
+				}
+			});
+		}
+
+		// Check for confirmation
+		if (!confirmation.confirmed) {
+			return false;
+		}
+
+		// Check for confirmation checkbox
+		if (confirmation.confirmed && confirmation.checkboxChecked === true) {
+			await this._configurationService.updateValue(TerminalSettingId.MultiLinePasteWarning, false);
+		}
+		return true;
+	}
```

Comment:
Could be simplified:

```suggestion
		// Don't ask again if the checkbox is checked
		if (confirmation.confirmed && confirmation.checkboxChecked) {
			await this._configurationService.updateValue(TerminalSettingId.MultiLinePasteWarning, false);
		}
		
		return confirmation.confirmed;
	}
```